### ISSUE TITLE
[Bug]:  Cannot locate specified Dockerfile: Dockerfile, will retry 

### ISSUE THREAD

**Issue by danvergara** [testcontainers/testcontainers-go#2203]
### Testcontainers version

0.27.0

### Using the latest Testcontainers version?

Yes

### Host OS

Debian

### Host arch

x86

### Go version

1.21.6

### Docker version

```shell
Client: Docker Engine - Community
 Version:           25.0.1
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Tue Jan 23 23:09:46 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          25.0.1
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       71fa3ab
  Built:            Tue Jan 23 23:09:46 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.27
  GitCommit:        a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc:
  Version:          1.1.11
  GitCommit:        v1.1.11-0-g4bccb38
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
Client: Docker Engine - Community
 Version:    25.0.1
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 4
  Running: 3
  Paused: 0
  Stopped: 1
 Images: 42
 Server Version: 25.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc version: v1.1.11-0-g4bccb38
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-5-amd64
 Operating System: Debian GNU/Linux trixie/sid
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.13GiB
 Name: danvergara
 ID: 907647ef-4861-4c09-bdc8-0449fe6f2bad
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: fals
```

### What happened?

I need to build a container based off the Dockerfile that is located at the root at the project. So, every time I run this something like the example below I get an error like the title.

```go
c, err := GenericContainer(ctx, GenericContainerRequest{
    ContainerRequest: ContainerRequest{
        FromDockerfile: FromDockerfile{
            Context:       ".",
            PrintBuildLog: true,
            BuildOptionsModifier: func(buildOptions *types.ImageBuildOptions) {
                buildOptions.Target = "development"
            },
        },
    },
    Started: true,
})
```

### Relevant log output

```shell
2024/02/06 21:35:16 Failed to build image: error during connect: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.44/build?buildargs=%7B%22GITLAB_TOKEN%22%3A%22%22%7D&cachefrom=null&cgroupparent=&cpuperiod=0&cpuquota=
0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile&forcerm=1&labels=null&memory=0&memswap=0&networkmode=&rm=1&shmsize=0&t=71107202-5f53-46db-8004-53ef95c17601%3A48df2665-8859-4ad2-ac21-e1974f6b2f2e&targ
et=development&ulimits=null&version=": io: read/write on closed pipe, will retry
```

**Comment by danvergara** [comment]
Update: I spotted the root issue and that is that the `Dockerfile` in the `.dockerignore` file. Don't know if this is expected or is unintended, since I wouldn't add the `Dockerfile` to the build process.

**Comment by mdelapenya** [comment]
I've done a simple test:

0. create a new directory and move there

```shell
mkdir /tmp/ignoredockerfile
cd /tmp/ignoredockerfile
```

1. create a `.dockerignore` file:
```
Dockerfile
```

2. create a Dockefile

```docker
FROM alpine
RUN echo 'foo'
```

3. build the image

```shell
docker build -t ignoreme .
[+] Building 0.2s (6/6) FINISHED                                                                                                                                                                                        docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                                                                    0.0s
 => => transferring dockerfile: 64B                                                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                                                                                                        0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                       0.0s
 => => transferring context: 51B                                                                                                                                                                                                        0.0s
 => [1/2] FROM docker.io/library/alpine:latest                                                                                                                                                                                          0.0s
 => [2/2] RUN echo 'foo'                                                                                                                                                                                                                0.1s
 => exporting to image                                                                                                                                                                                                                  0.0s
 => => exporting layers                                                                                                                                                                                                                 0.0s
 => => writing image sha256:26964df416e0ddb33d9d12f6f0ce47227affbb46b19ee045970910643d52e07a                                                                                                                                            0.0s
 => => naming to docker.io/library/ignoreme                                                                                                                                                                                             0.0s

What's Next?
  View a summary of image vulnerabilities and recommendations → docker scout quickview
```

The image is built, so I think the library should honour this behaviour.

Would you be interested in contributing the fix?

**Comment by danvergara** [comment]
Yeah, I can fix that!
Thanks @mdelapenya

**Comment by molteninjabob** [comment]
@danvergara I'm new to test containers and I think I just ran into this same issue. I'm trying to create a container using a Dockerfile in a different directory, but I've confirmed the Dockerfile is in the directory being referenced from Context. I noticed your thread in the Slack workspace about .dockerignore and confirmed we're doing the same thing.

Do you have an ETA on a fix? Is there a workaround for this?

**Comment by danvergara** [comment]
@molteninjabob yeah I have an idea on what's going on under the hood, but I haven been very busy with my daily job. I'll get it fixed as soon as I can.

**Comment by danvergara** [comment]
I managed to reproduce the error in the test suite. I just added a `dockerignore` file with a couple of `Dockerfiles` in it and some tests failed, as expected. Working on a fix rn.

![Screenshot from 2024-02-28 09-57-08](https://github.com/testcontainers/testcontainers-go/assets/12239167/4cfe474e-c20d-424d-a8c4-5bddb23f1c9a)
![Screenshot from 2024-02-28 09-59-21](https://github.com/testcontainers/testcontainers-go/assets/12239167/d0223bea-a52a-4dd9-b0f5-69fcb341c777)
